src/libexpr/primops.cc

```diff
@@ -1043,6 +1043,49 @@ static RegisterPrimOp primop_trace({
     .fun = prim_trace,
 });
 
+static void prim_warn(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    // We only accept a string argument for now. The use case for pretty printing a value is covered by `trace`.
+    // By rejecting non-strings we allow future versions to add more features without breaking existing code.
+    auto msgStr = state.forceString(*args[0], pos, "while evaluating the first argument; the message passed to builtins.warn");
+
+    {
+        BaseError msg(std::string{msgStr});
+        msg.atPos(state.positions[pos]);
+        auto info = msg.info();
+        info.level = lvlWarn;
+        logWarning(info);
+    }
+
+    if (evalSettings.builtinsAbortOnWarn) {
+        state.error<Abort>("aborting to reveal stack trace of warning, as abort-on-warn is set").debugThrow();
+    }
+    if ((evalSettings.builtinsTraceDebugger || evalSettings.builtinsDebuggerOnWarn) && state.debugRepl && !state.debugTraces.empty()) {
+        const DebugTrace & last = state.debugTraces.front();
+        state.runDebugRepl(nullptr, last.env, last.expr);
+    }
```

Comment:
A few of the call sites `runDebugRepl` look similar --- not just `builtin.trace`'s even. Should we abstract over this a bit?

Reply:
I think yes, by allowing any primop to be marked as a breakpoint.
Until we have that, let's just keep it simple.

Reply:
Just to be clear, I mean not a new generalized feature, but something very mechanical:
```c++
if (state.debugRepl && !state.debugTraces.empty()) {
    const DebugTrace & last = state.debugTraces.front();
    state.runDebugRepl(<some-expression>, last.env, last.expr);
}
```
seems to be part of this `EvalErrorBuilder<T>::debugThrow`, `primop_break`, `primop_trace`, and this. So I am wondering if a new function (maybe in header, taking an `auto &&` lambda, to get the laziness right) is warranted.